client/finality-grandpa/src/authorities.rs

```diff
@@ -454,13 +503,16 @@ where
 					"block" => ?change.canon_height
 				);
 
+				self.authority_set_changes.append(self.set_id, median_last_finalized.clone());
```

Comment:
This method does not mutate and instead returns a new `AuthoritySet` if there are any changes. This is how we handle rollbacks in case we apply a change and later on in the process the block being imported fails for some reason. Instead we should keep this method immutable and we should instead create the appended version of `AuthoritySetChanges` and use it below.